shell/browser/api/electron_api_web_frame_main.cc

```diff
@@ -0,0 +1,230 @@
+#include "content/public/browser/render_frame_host.h"
+#include "gin/object_template_builder.h"
+#include "shell/browser/browser.h"
+#include "shell/common/gin_converters/frame_converter.h"
+#include "shell/common/gin_converters/gurl_converter.h"
+#include "shell/common/gin_converters/value_converter.h"
+#include "shell/common/gin_helper/dictionary.h"
+#include "shell/common/gin_helper/error_thrower.h"
+#include "shell/common/gin_helper/object_template_builder.h"
+#include "shell/common/gin_helper/promise.h"
+#include "shell/common/node_includes.h"
+
+namespace electron {
+
+namespace api {
+
+typedef std::unordered_map<content::RenderFrameHost*, WebFrame*> RenderFrameMap;
+base::LazyInstance<RenderFrameMap>::DestructorAtExit g_render_frame_map =
+    LAZY_INSTANCE_INITIALIZER;
+
+gin::WrapperInfo WebFrame::kWrapperInfo = {gin::kEmbedderNativeGin};
+
+// WebFrame can outlive its RenderFrameHost pointer so we need to check whether
+// its been disposed of prior to accessing it.
+#define CHECK_RENDER_FRAME(funcname, args, value)                      \
+  if (render_frame_disposed_) {                                        \
+    args->isolate()->ThrowException(v8::Exception::Error(              \
+        gin::StringToV8(args->isolate(),                               \
+                        "Render frame was disposed before "            \
+                        "WebFrame." #funcname " could be executed"))); \
+    return value;                                                      \
+  }
+
+WebFrame::WebFrame(content::RenderFrameHost* rfh) : render_frame_(rfh) {
+  g_render_frame_map.Get().emplace(rfh, this);
+}
+
+WebFrame::~WebFrame() {
+  MarkRenderFrameDisposed();
+}
+
+void WebFrame::MarkRenderFrameDisposed() {
+  g_render_frame_map.Get().erase(render_frame_);
+  render_frame_disposed_ = true;
+}
+
+v8::Local<v8::Promise> WebFrame::ExecuteJavaScript(const base::string16& code,
+                                                   bool has_user_gesture,
+                                                   gin::Arguments* args) {
+  gin_helper::Promise<base::Value> promise(args->isolate());
+  v8::Local<v8::Promise> handle = promise.GetHandle();
+
+  if (render_frame_disposed_) {
+    promise.RejectWithErrorMessage(
+        "Render frame was disposed before WebFrame.executeJavaScript could be "
+        "executed");
+    return handle;
+  }
+
+  if (has_user_gesture) {
+    auto* ftn = content::FrameTreeNode::From(render_frame_);
+    ftn->UpdateUserActivationState(
+        blink::mojom::UserActivationUpdateType::kNotifyActivation,
+        blink::mojom::UserActivationNotificationType::kTest);
+  }
+
+  render_frame_->ExecuteJavaScriptForTests(
+      code, base::BindOnce([](gin_helper::Promise<base::Value> promise,
+                              base::Value value) { promise.Resolve(value); },
+                           std::move(promise)));
+
+  return handle;
+}
+
+bool WebFrame::Reload(gin::Arguments* args) {
```

Comment:
These functions can all take `v8::Isolate*` instead of `gin::Arguments*`, which will cause them to correctly throw errors when they receive more arguments than expected.